File: src/main/cpp/commands/command_DriveAuton.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "commands/command_DriveAuton.h"
#include "frc/smartdashboard/SmartDashboard.h"

command_DriveAuton::command_DriveAuton(subsystem_DriveTrain* DriveTrain, subsystem_PoseTracker* PoseTracker, std::string TrajFilePath, bool ToReset):
m_DriveTrain{DriveTrain}, m_PoseTracker{PoseTracker}, m_ToReset{ToReset},  m_DriveController{AutoConstants::XPID, AutoConstants::YPID, AutoConstants::ThetaPID} {
  // pathplanner::PPHolonomicDriveController m_DriveController{AutoConstants::XPID, AutoConstants::YPID, AutoConstants::ZPID};
  // Use addRequirements() here to declare subsystem dependencies.
  AddRequirements({m_DriveTrain});
  AddRequirements({m_PoseTracker});
  // m_Trajectory = frc::TrajectoryUtil::FromPathweaverJson(deployDirectory.string());

  if(TrajFilePath.starts_with("Yes^3") ){
    m_Trajectory = pathplanner::PathPlanner::loadPath(TrajFilePath, pathplanner::PathConstraints(4_mps, units::meters_per_second_squared_t{4}) );  

  }else{
    m_Trajectory = pathplanner::PathPlanner::loadPath(TrajFilePath, pathplanner::PathConstraints(AutoConstants::MaxSpeed, AutoConstants::MaxAccel) );  

  }






// command_DriveAuton::command_DriveAuton(subsystem_DriveTrain* DriveTrain, subsystem_PoseTracker* PoseTracker, std::string TrajFilePath, bool ToReset):
// m_DriveTrain{DriveTrain}, m_PoseTracker{PoseTracker}, m_ToReset{ToReset}, m_IsCustomMotion{false}, m_DriveController{AutoConstants::XPID, AutoConstants::YPID, AutoConstants::ThetaPID} {
  // pathplanner::PPHolonomicDriveController m_DriveController{AutoConstants::XPID, AutoConstants::YPID, AutoConstants::ZPID};
  // Use addRequirements() here to declare subsystem dependencies.
  // AddRequirements({m_DriveTrain});
  // AddRequirements({m_PoseTracker});
  // m_Trajectory = frc::TrajectoryUtil::FromPathweaverJson(deployDirectory.string());


  // m_Trajectory = pathplanner::PathPlanner::loadPath(TrajFilePath, pathplanner::PathConstraints(AutoConstants::MaxSpeed, AutoConstants::MaxAccel) );  
  
  
}

// Called when the command is initially scheduled.
void command_DriveAuton::Initialize() {
  m_Trajectory = pathplanner::PathPlannerTrajectory::transformTrajectoryForAlliance(m_Trajectory, frc::DriverStation::GetAlliance());
  
  m_Timer.Stop();
  m_Timer.Reset();
  m_Timer.Start();

  if( m_ToReset ){
    m_DriveTrain->ResetOdometry(m_Trajectory.getInitialHolonomicPose());
  }
  frc::SmartDashboard::GetNumber("InitialX", m_Trajectory.getInitialHolonomicPose().X().value());
  frc::SmartDashboard::GetNumber("InitialY", m_Trajectory.getInitialHolonomicPose().Y().value());
  frc::SmartDashboard::GetNumber("InitialHeading", m_Trajectory.getInitialHolonomicPose().Rotation().Degrees().value());

  
}

// Called repeatedly when this Command is scheduled to run
void command_DriveAuton::Execute() {
  // if(m_DriveTrain->HasAcceptableTargets()){
  //   m_DriveTrain->ImplementVisionPose(m_DriveTrain->getEstimatedGlobalPose());    
  // }
  // m_DriveTrain->ImplementVisionPose(m_PoseTracker->GetEstimatedGlobalPose());

  pathplanner::PathPlannerTrajectory::PathPlannerState state = m_Trajectory.sample(m_Timer.Get()); 
  auto chassisSpeeds = m_DriveController.Calculate(m_DriveTrain->GetPose(), 
                                                  state.pose,
                                                  state.velocity,
                                                  state.holonomicRotation);
  // chassisSpeeds.vy = units::meters_per_second_t{ fabs(chassisSpeeds.vy.value()) };
  auto ModuleStates = SwerveConstants::m_kinematics.ToSwerveModuleStates(chassisSpeeds);
  SwerveConstants::m_kinematics.DesaturateWheelSpeeds(&ModuleStates, AutoConstants::MaxSpeed);
  m_DriveTrain->SetModuleStates(ModuleStates);

  
}

// Called once the command ends or is interrupted.
void command_DriveAuton::End(bool interrupted) {
  m_DriveTrain -> SwerveDrive(0_mps, 0_mps, 0_rad_per_s, false, false);
}

// Returns true when the command should end.
bool command_DriveAuton::IsFinished() {
  return m_Timer.Get() >= (m_Trajectory.getTotalTime());
}

File: src/main/java/frc/robot/commands/command_DriveAuton.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import java.util.function.BooleanSupplier;
// import java.util.function.DoubleSupplier;

import com.pathplanner.lib.path.PathPlannerPath;
import com.pathplanner.lib.path.PathPlannerTrajectory;

import edu.wpi.first.math.controller.HolonomicDriveController;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.CommandBase;
// import frc.robot.SwerveModule;
import frc.robot.Constants.AutoConstants;
import frc.robot.Constants.SwerveConstants;
import frc.robot.subsystems.subsystem_DriveTrain;

public class command_DriveAuton extends CommandBase {
  /** Creates a new command_DriveAuton. */
  private subsystem_DriveTrain m_DriveTrain;
  private BooleanSupplier m_ToReset;
  private Timer m_Timer;

  private PathPlannerTrajectory m_Trajectory;
  private PathPlannerPath m_Path;
  private HolonomicDriveController m_DriveController;

  public command_DriveAuton(subsystem_DriveTrain driveTrain,
                            BooleanSupplier toReset,
                            String trajFilePath) {
    // Use addRequirements() here to declare subsystem dependencies.
    m_DriveTrain = driveTrain;
    m_ToReset = toReset;
    addRequirements(m_DriveTrain);

    m_Path = PathPlannerPath.fromPathFile(trajFilePath);
    m_Trajectory = new PathPlannerTrajectory(m_Path, new ChassisSpeeds(0, 0, 0));
    m_Timer = new Timer();
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {
    m_Timer.stop();
    m_Timer.reset();
    m_Timer.start();

    Pose2d initOdometry = m_Trajectory.getInitialTargetHolonomicPose();

    if(m_ToReset.getAsBoolean()){
      m_DriveTrain.resetOdometry(initOdometry);
    }

    SmartDashboard.getNumber("Initial_X", initOdometry.getX());
    SmartDashboard.getNumber("Initial_Y", initOdometry.getX());
    SmartDashboard.getNumber("Initial_Heading", initOdometry.getRotation().getDegrees());
  }

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {

    PathPlannerTrajectory.State state = m_Trajectory.sample(m_Timer.get());
    ChassisSpeeds chassisSpeeds = m_DriveController.calculate(m_DriveTrain.getPose(),
                                              state.getTargetHolonomicPose(),
                                              state.velocityMps,
                                              state.targetHolonomicRotation);
    SwerveModuleState[] moduleStates = SwerveConstants.kinematics.toSwerveModuleStates(chassisSpeeds);
    SwerveDriveKinematics.desaturateWheelSpeeds(moduleStates, AutoConstants.MaxSpeedMetersPerSecond);
    m_DriveTrain.setModuleStates(moduleStates);
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {
    m_DriveTrain.swerveDrive(0, 0, 0, false, false);
  }

  // Returns true when the command should end.
  @Override
  public boolean isFinished() {
    return m_Timer.get() >= m_Trajectory.getTotalTimeSeconds();
  }
}
